### PR TITLE
[WIP] add request signature

### PR DESCRIPTION
**Pull request by jabbink**
WIP

Todo
- [x] call encrypt
- [x] move signature stuff to different class

### CODE REVIEW

**Comment by Paul776** [comment]
Nice, How will you calculate the sig?

**Comment by jay8ee** [comment]
Are you going to try and generate it natively?

**Comment by jabbink** [comment]
It does that right now. Here's a Java impl https://gist.github.com/jabbink/213e22c12db4273f0e7161eec014d327

**Comment by jabbink** [comment]
But I have to go now, so feel free to change this PR/create a new one.

Shouldn't be too much work with the code in the changed files.

**Comment by Paul776** [comment]
Possible bug: nuilder.getAuthTicket() always empty
it is never et anymore in ur PR

**Comment by jabbink** [comment]
Fixed in the previous commit afaik

**Comment by Paul776** [comment]
Just tried (with JNA imp) still get 0 to everything in the Map object
(python scripts still works)

**Comment by Menes1337** [comment]
com/pokegoapi/util/Crypto.java:
`import javax.xml.bind.DatatypeConverter;`

seems not available in Android. I got an exception(java.lang.UnsatisfiedLinkError: Native library..) I think you just use it for debugging purpose. Thats why I just comment the lines out:

`//String hex = DatatypeConverter.printHexBinary(t);
//System.out.println(hex.toLowerCase());`

**Comment by Paul776** [comment]
Something is wrong in getLocationHash1 you need to seed it with the ticket hash

**Comment by vmarchaud** [comment]
@Paul776 yes already told him and got fixed in last commit.

File: library/src/main/java/com/pokegoapi/main/RequestHandler.java [hunk 1]
```diff
@@ -313,4 +302,77 @@ public void run() {
...
+		}
+
+		// TODO: Call encrypt function on this
+		ByteString uk2 = sigBuilder.build().toByteString();
+		Unknown6OuterClass.Unknown6.newBuilder()
+				.setRequestType(6)
+				.setUnknown2(Unknown6OuterClass.Unknown6.Unknown2.newBuilder().setUnknown1(uk2));
+	}
+
+	private int getLocationHash1(byte[] authTicket, RequestEnvelope.Builder builder) {
```
Comment (Paul776) [inline library/src/main/java/com/pokegoapi/main/RequestHandler.java]: There is an other problem here, something is not right with the minus sign of the longitude

Comment (Paul776) [inline library/src/main/java/com/pokegoapi/main/RequestHandler.java]: I don't get the same result as if I use the original python script

Comment (jabbink) [inline library/src/main/java/com/pokegoapi/main/RequestHandler.java]: java uses signed int's. The python code unsigned.

Try to dump them both as hexadecimal (Integer.toHexString() in Java):

what is sent is exactly the same.

File: library/src/main/java/com/pokegoapi/main/RequestHandler.java [hunk 2]
```diff
@@ -313,4 +304,94 @@ public void run() {
...
+		XXHashFactory factory = XXHashFactory.fastestInstance();
+		StreamingXXHash32 xx32 = factory.newStreamingHash32(0x1B845238);
+		xx32.update(authTicket, 0, authTicket.length);
+		xx32 = factory.newStreamingHash32(xx32.getValue());
+		byte[] bytes = new byte[8 * 3];
+
+		System.arraycopy(getBytes(api.getLatitude()), 0, bytes, 0, 8);
+		System.arraycopy(getBytes(api.getLongitude()), 0, bytes, 8, 8);
```
Comment (Paul776) [inline library/src/main/java/com/pokegoapi/main/RequestHandler.java]: Something is not right here with the minus sign of the location
if the python script xx32 hashed this -0.1277583 (with the 0x1B845238) it returns: 2878327542
here it return -1416639754

x32.hash(-0.1277583), 0, getBytes(-0.1277583).length, 0x1B845238)

(this is just a test to verify that the hashes are correct, as am still getting 0 pokemons back)

Comment (jabbink) [inline library/src/main/java/com/pokegoapi/main/RequestHandler.java]: As it is sent via protobuf I think that shouldn't matter, as it's the same memory representation

http://ideone.com/Y0DLIo

Comment (Paul776) [inline library/src/main/java/com/pokegoapi/main/RequestHandler.java]: I see.. something is not right somewhere then. Is it just me not getting anything back? Or is it working for you ?

Comment (jabbink) [inline library/src/main/java/com/pokegoapi/main/RequestHandler.java]: @vmarchaud has the solution, he says. This code is not working for anyone, as I've said in the comments (and this PR isn't merged, and it's called WIP).

Comment (iGio90) [inline library/src/main/java/com/pokegoapi/main/RequestHandler.java]: Im getting valid response with @jabbink impl btwn! WP

**Comment by iGio90** [comment]
Im getting valid response! between, do you know if the other messages inside the signature (device infos etc) are also encr or not?

**Comment by vmarchaud** [comment]
@iGio90 Currently nobody is sending them, they might force us to put them in the following days, we will see how to implement this.

**Comment by iGio90** [comment]
@vmarchaud i saw your implementation today and i think its better to include these info directly in the impl. The UDID you used could be simply replaced with an Alphanum string of 40 lenght. All the other things you used are randomed good. It would be simple! do you know if they are encrypted or normal proto? i can make some test and PR in case!

**Comment by vmarchaud** [comment]
@iGio90 removed it in the release, it was just to test last night. For the high level API we need to think a impl where user can put its own faked data.

**Comment by Paul776** [comment]
http://imgur.com/a/Ez9yt

**Comment by jabbink** [comment]
:tada:

**Comment by jay8ee** [comment]
How can I actually get these changes and start testing? Can't seem to compile against the development branch with gradle.

**Comment by jabbink** [comment]
Now it should be on jcenter

**Comment by iGio90** [comment]
@vmarchaud could you please explain me why we should let the users insert data instead of sending fake ones? the purpose is just to prevent some sort of flags on missing device id. And also, 2 line serverside on device id will stop everything again! Still can't get why we couldn't send fake data

**Comment by vmarchaud** [comment]
@iGio90 cause the problem is the same in your version, Niantic just has to ban ONE fake data to ban all our users, come on slack if you have more question.

**Comment by iGio90** [comment]
@vmarchaud coming now! (edit: teamdomain?)

**Comment by vmarchaud** [comment]
@iGio90 everything is at the bottom of the readme

File: library/src/main/java/com/pokegoapi/util/Signature.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+				.setTimestamp(api.currentTimeMillis())
+				.setTimestampSinceStart(curTime - api.startTime);
+
+
+		for (RequestOuterClass.Request serverRequest : builder.getRequestsList()) {
+			byte[] request = serverRequest.toByteArray();
+			sigBuilder.addRequestHash(getRequestHash(authTicketBA, request));
+		}
+
+		// TODO: Call encrypt function on this
```
Comment (ScottPierce) [inline library/src/main/java/com/pokegoapi/util/Signature.java:51]: @jabbink So what's to happen with this TODO?

**Pull request merged**